### PR TITLE
Add ignore:["custom-element"] option to selector-type-no-unknown

### PR DESCRIPTION
**Pull request by evilebottnawi**
> Which issue, if any, is this issue related to?

#2348

> Is there anything in the PR that needs further explanation?

No, it's self explanatory.

Btw, i think be good add option `resolve: ['custom-element']` (maybe best name). Because `not-my-type` is valid as custom element but invalid as known type selector :sob:

### CODE REVIEW

**Comment by jeddy3** [comment]
Can you update [the docs](https://github.com/stylelint/stylelint/tree/master/lib/rules/selector-type-no-unknown#optional-secondary-options):

> This rule considers tags defined in the HTML and SVG Specifications to be known.

Becomes:

This rule considers tags which are [valid custom elements](https://www.w3.org/TR/custom-elements/#valid-custom-element-name) and those defined in the following Specifications to be known: HTML and SVG.

(We'll [likely be adding](https://github.com/stylelint/stylelint/issues/2326#issuecomment-278416740) MathML to that list).

Can you also update the [examples](https://github.com/stylelint/stylelint/tree/master/lib/rules/selector-type-no-unknown#ignoretypes-regex-string) not to use hyphens.

Lastly, can you remove the hyphen from [the regex tests](https://github.com/stylelint/stylelint/blob/80ed0c7200047bcd47476e7a56e74330c0c34ea4/lib/rules/selector-type-no-unknown/__tests__/index.js#L173-L198) to make them relevant again.

> Btw, i think be good add option resolve: ['custom-element'] (maybe best name). Because not-my-type is valid as custom element but invalid as known type selector

Can you reword this please? It's not clear what you mean.

**Comment by evilebottnawi** [comment]
@jeddy3 if we merge this PR type selector `foo-bar` be valid type selector (because it is valid as custom selector), if we want avoid this situation we should add option `resolve: ["custom-selector"]` (maybe best), if option enable `foo-bar` is valid as type selector, if option disable `foo-bar` is invalid as type selector, sorry for my bad english

**Comment by jeddy3** [comment]
Oh I see. You're suggesting we put this behind an option e.g. `ignore: ["custom-elements"]`?

>  if option enable foo-bar is valid as type selector, if option disable foo-bar is invalid as type selector

That would address this issue.

**Comment by evilebottnawi** [comment]
@jeddy3 yep, it will be good solution

**Comment by davidtheclark** [comment]
I'm wondering if this option should be reversed. Instead of by default complaining about custom elements, it seems we should by default let them pass. Maybe we'd add a `disallowCustomElements` option to get the reverse.

(It's probably unlikely that a legitimate custom element is going to be confused with a type selector typo, because custom elements require hyphens and almost no real type selectors do.)

**Comment by jeddy3** [comment]
> I'm wondering if this option should be reversed. Instead of by default complaining about custom elements, it seems we should by default let them pass.

I agree.

> Maybe we'd add a disallowCustomElements option to get the reverse.

Yep, but I think we can wait until someone actually runs into the scenario and requests it (as, like you said, it seems unlikely to occur).

So, I think this PR is ready to merge once @evilebottnawi addresses the [doc and test requests](https://github.com/stylelint/stylelint/pull/2366#issuecomment-280470173).

**Comment by evilebottnawi** [comment]
@jeddy3 some my project use custom elements some no, i am add this option :smile: , i requests this :smile:

**Comment by jeddy3** [comment]
> some my project use custom elements some no, i am add this option 😄 , i requests this

Sounds good. Let's create a separate issue for that and add that option in a separate PR. Let's handle the [doc and test requests](https://github.com/stylelint/stylelint/pull/2366#issuecomment-280470173) and merge this.

**Comment by davidtheclark** [comment]
@evilebottnawi Do you plan to finish this PR, or should we close it?

**Comment by evilebottnawi** [comment]
@davidtheclark finish on this week :smile:

File: lib/utils/isCustomElement.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+const svgTags = require("svg-tags")
+const htmlTags = require("html-tags")
+const keywordSets = require("../reference/keywordSets")
+const mathmlTagNames = require("mathml-tag-names")
+
+/**
+ * Check whether a type selector is a custom element
+ */
+module.exports = function (selector/*: string*/)/*: boolean*/ {
+  const startWithALowercaseASCII = (/^[a-z]/m).test(selector)
```
Comment (davidtheclark) [inline lib/utils/isCustomElement.js]: Why the `m` flag?

File: lib/utils/isCustomElement.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  const isNotHtmlTag = htmlTags.indexOf(selector.toLowerCase()) === -1
+    && !keywordSets.nonStandardHtmlTags.has(selector.toLowerCase())
+  const isNotMathmlTag = mathmlTagNames.indexOf(selector.toLowerCase()) === -1
+
+  return startWithALowercaseASCII
+    && notContainAnyUppercaseASCII
+    && containAHyphen
+    && isNotSvgTag
+    && isNotHtmlTag
+    && isNotMathmlTag
```
Comment (davidtheclark) [inline lib/utils/isCustomElement.js]: This is kind of a weird way to structure the function. Couldn't you have a return after each statement and avoid the need to assign variables? E.g.

```js
if (!(/^[a-z]/m).test(selector)) return false
if (selector.toLowerCase() !== selector) return false
...
```

**Comment by jeddy3** [comment]
> I'm wondering if this option should be reversed. Instead of by default complaining about custom elements, it seems we should by default let them pass. Maybe we'd add a `disallowCustomElements` option to get the reverse.

I just realised, while thinking through https://github.com/kristerkari/stylelint-scss/issues/103#issuecomment-287649576, that perhaps this behaviour should be behind an option. Sorry to flip-flop.

I believe _every_ `no` rule in stylelint is most restrictive by default and can be made more permissive using `"ignore": []` options. We even changed `indentation` to be consistent with this behaviour a while back, but checking function parameters _by default_ rather than via opt-in. 

I think adding a `disallow...` boolean option would be at odds with what we've done elsewhere... I think it's also the reason https://github.com/stylelint/stylelint/issues/2437 confused me - that is the issue for this `disallow...` option, right? But it looks like @evilebottnawi instinctively, and understandably so, reach for an `ignore*` option. 

Putting this behind an option also fits with the idea of "known" being a discrete list of elements from specifications, whereas `ignore: ["custom-elements"]` is triggering a _heuristic_ ignore. 

Want do you think?

**Comment by evilebottnawi** [comment]
@jeddy3 sound good for option :+1:

**Comment by jeddy3** [comment]
> sound good for option

@evilebottnawi Is this change something you can factor into this PR?

**Comment by evilebottnawi** [comment]
@jeddy3 yep, in near future

File: lib/utils/isCustomElement.js (hunk 1 above)
Comment (evilebottnawi) [inline lib/utils/isCustomElement.js]: @davidtheclark yep, remove it

File: lib/utils/isCustomElement.js (hunk 2 above)
Comment (evilebottnawi) [inline lib/utils/isCustomElement.js]: @davidtheclark done

File: lib/rules/selector-type-no-unknown/index.js [hunk 3]
```diff
@@ -97,10 +74,20 @@ const rule = function (actual, options) {
             return
           }
 
+          if (
```
Comment (jeddy3) [inline lib/rules/selector-type-no-unknown/index.js]: Minor request... can we get this shifted to [line 61](https://github.com/stylelint/stylelint/blob/fff628f10cabf424d18c79750c7e5f71a819ad50/lib/rules/selector-type-no-unknown/index.js#L61) so the early returns are in the same alphabetical order as the options config.

Quite a lot of rules now have 5 or more `ignore/except*` options and I think there's no harm adding a bit of consistent structure to them.

File: lib/rules/selector-type-no-unknown/index.js [hunk 4]
```diff
@@ -19,39 +21,14 @@ const messages = ruleMessages(ruleName, {
...
-])
-
 const rule = function (actual, options) {
   return (root, result) => {
     const validOptions = validateOptions(result, ruleName, { actual }, {
       actual: options,
       possible: {
         ignore: [
           "default-namespace",
+          "custom-elements",
```
Comment (jeddy3) [inline lib/rules/selector-type-no-unknown/index.js]: Alphabetical order please.

File: lib/rules/selector-type-no-unknown/README.md [hunk 5]
```diff
@@ -8,7 +8,7 @@ Disallow unknown type selectors.
  * This type selector */
 ```
 
-This rule considers tags defined in the HTML and SVG Specifications to be known.
+This rule considers tags which are valid custom elements and those defined in the following Specifications to be known: HTML and SVG.
```
Comment (jeddy3) [inline lib/rules/selector-type-no-unknown/README.md]: This rule considers tags which are defined in the following Specifications to be known: HTML and SVG.

**Comment by awebdeveloper** [comment]
is this merged. Can a new tag be created

**Comment by evilebottnawi** [comment]
@awebdeveloper wait after @davidtheclark accept.

**Comment by awebdeveloper** [comment]
@evilebottnawi just to clarify will this work for all the below cases

```
1. my-element
2. element
3. [my-attr]
```

**Comment by davidtheclark** [comment]
Added to changelog:

> -   Fixed: `selector-type-no-unknown` now ignores valid custom elements ([#2366](https://github.com/stylelint/stylelint/pull/2366)).

**Comment by abdonrd** [comment]
Thanks! 👏  🎉

**Comment by awebdeveloper** [comment]
looking at the docs this doesn't take into account custom attributes. **[my-attr]**. Maybe a different PR but that should also be included

**Comment by evilebottnawi** [comment]
@awebdeveloper please create issue with template for this

**Pull request merged**